File: app/src/main/java/com/example/data/NavigationStackState.kt
package com.example.data

import androidx.compose.runtime.mutableStateListOf
import kotlin.reflect.KProperty

class NavigationStackState(initial: String) {
    private val stack = mutableStateListOf(initial)

    val canGoBack: Boolean
        get() = stack.size > 1

    fun goBack() {
        if (stack.size > 1) {
            stack.removeAt(stack.lastIndex)
        }
    }

    operator fun getValue(thisRef: Any?, property: KProperty<*>): String = stack.last()

    operator fun setValue(thisRef: Any?, property: KProperty<*>, value: String) {
        if (stack.lastOrNull() != value) {
            stack.add(value)
        }
    }
}

File: NavigationStackState.swift
import Foundation
import Combine

class NavigationStackState: ObservableObject {
    @Published private var stack: [String]
    
    init(initial: String) {
        self.stack = [initial]
    }
    
    var current: String {
        get { stack.last ?? "" }
        set {
            if stack.last != newValue {
                stack.append(newValue)
            }
        }
    }
    
    var canGoBack: Bool {
        stack.count > 1
    }
    
    func goBack() {
        if stack.count > 1 {
            stack.removeLast()
        }
    }
}
